File: MPChartLib/src/main/java/com/github/mikephil/charting/buffer/HorizontalBarBuffer.java

package com.github.mikephil.charting.buffer;

import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;

public class HorizontalBarBuffer extends BarBuffer {

    public HorizontalBarBuffer(int size, float groupspace, int dataSetCount, boolean containsStacks) {
        super(size, groupspace, dataSetCount, containsStacks);
    }

    @Override
    public void feed(IBarDataSet data) {

        float size = data.getEntryCount() * phaseX;
        float barWidthHalf = mBarWidth / 2f;

        for (int i = 0; i < size; i++) {

            BarEntry e = data.getEntryForIndex(i);

            if(e == null)
                continue;

            float x = e.getX();
            float y = e.getY();
            float[] vals = e.getYVals();

            if (!mContainsStacks || vals == null) {

                float top = x - barWidthHalf;
                float bottom = x + barWidthHalf;
                float left, right;
                if (mInverted) {
                    left = y >= 0 ? y : 0;
                    right = y <= 0 ? y : 0;
                } else {
                    right = y >= 0 ? y : 0;
                    left = y <= 0 ? y : 0;
                }

                // multiply the height of the rect with the phase
                if (right > 0)
                    right *= phaseY;
                else
                    left *= phaseY;

                addBar(left, top, right, bottom);

            } else {

                float posY = 0f;
                float negY = -e.getNegativeSum();
                float yStart = 0f;

                // fill the stack
                for (int k = 0; k < vals.length; k++) {

                    float value = vals[k];

                    if (value >= 0f) {
                        y = posY;
                        yStart = posY + value;
                        posY = yStart;
                    } else {
                        y = negY;
                        yStart = negY + Math.abs(value);
                        negY += Math.abs(value);
                    }

                    float top = x - barWidthHalf;
                    float bottom = x + barWidthHalf;
                    float left, right;
                    if (mInverted) {
                        left = y >= yStart ? y : yStart;
                        right = y <= yStart ? y : yStart;
                    } else {
                        right = y >= yStart ? y : yStart;
                        left = y <= yStart ? y : yStart;
                    }

                    // multiply the height of the rect with the phase
                    right *= phaseY;
                    left *= phaseY;

                    addBar(left, top, right, bottom);
                }
            }
        }

        reset();
    }
}

File: MPChartLib/src/main/java/com/github/mikephil/charting/buffer/HorizontalBarBuffer.kt
package com.github.mikephil.charting.buffer

import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import kotlin.math.abs

class HorizontalBarBuffer(size: Int, dataSetCount: Int, containsStacks: Boolean) : BarBuffer(size, dataSetCount, containsStacks) {
    override fun feed(data: IBarDataSet?) {
        val size = (data?.entryCount ?: 0) * phaseX
        val barWidthHalf = barWidth / 2f

        var i = 0
        while (i < size) {
            val e = data?.getEntryForIndex(i)

            if (e == null) {
                i++
                continue
            }

            val x = e.x
            var y = e.y
            val vals = e.yVals

            if (!containsStacks || vals == null) {
                val bottom = x - barWidthHalf
                val top = x + barWidthHalf
                var left: Float
                var right: Float
                if (inverted) {
                    left = if (y >= 0) y else 0f
                    right = if (y <= 0) y else 0f
                } else {
                    right = if (y >= 0) y else 0f
                    left = if (y <= 0) y else 0f
                }

                // multiply the height of the rect with the phase
                if (right > 0) right *= phaseY
                else left *= phaseY

                addBar(left, top, right, bottom)
            } else {
                var posY = 0f
                var negY = -e.negativeSum
                var yStart: Float

                // fill the stack
                for (k in vals.indices) {
                    val value = vals[k]

                    if (value >= 0f) {
                        y = posY
                        yStart = posY + value
                        posY = yStart
                    } else {
                        y = negY
                        yStart = (negY + abs(value.toDouble())).toFloat()
                        negY += abs(value.toDouble()).toFloat()
                    }

                    val bottom = x - barWidthHalf
                    val top = x + barWidthHalf
                    var left: Float
                    var right: Float
                    if (inverted) {
                        left = if (y >= yStart) y else yStart
                        right = if (y <= yStart) y else yStart
                    } else {
                        right = if (y >= yStart) y else yStart
                        left = if (y <= yStart) y else yStart
                    }

                    // multiply the height of the rect with the phase
                    right *= phaseY
                    left *= phaseY

                    addBar(left, top, right, bottom)
                }
            }
            i++
        }

        reset()
    }
}
